faucet.go

```diff
@@ -232,6 +118,50 @@ func cleanAndExpandPath(path string) string {
 	return filepath.Clean(os.ExpandEnv(path))
 }
 
+// Verify timelimit from determined IP and count how many time
+// its need to do a new action.
+func verifyTimeLimit(ip string, actionsTimeLimit time.Duration, mutex sync.RWMutex) error {
+	mutex.RLock()
+	lastRequestTime, found := requestIPs[ip]
+	mutex.RUnlock()
+	if found {
+		nextAllowedRequest := lastRequestTime.Add(actionsTimeLimit)
+		coolDownTime := time.Until(nextAllowedRequest)
+
+		if coolDownTime >= 0 {
+			err := fmt.Errorf("client(%v) may only do an action every "+
+				"%v. Wait another %v.",
+				ip, actionsTimeLimit, coolDownTime)
+			return err
+		}
+	}
+	return nil
+}
+
+// Get the client's real IP address
+func getRealIP(r *http.Request, useRealIP bool) (string, error) {
+	if useRealIP {
+		xForwardFor := http.CanonicalHeaderKey("X-Forwarded-For")
+		xRealIP := http.CanonicalHeaderKey("X-Real-IP")
+
+		if xrip := r.Header.Get(xRealIP); xrip != "" {
+			return xrip, nil
+		} else if xff := r.Header.Get(xForwardFor); xff != "" {
+			i := strings.Index(xff, ", ")
+			if i == -1 {
+				i = len(xff)
+			}
+			return xff[:i], nil
+		}
+	}
+	log.Warn(`"X-Forwarded-For" header invalid, using RemoteAddr instead`)
```

Comment:
The position of this logging seems wrong. If the user sets `useRealIP` config to false, we will log a warning on every single request. Maybe it should be moved up one line, and perhaps it should read like `"X-Forwarded-For" and "X-Real-IP" headers invalid, using RemoteAddr instead`